### PR TITLE
Eliminate ADBinaries, add ADSupport

### PR DESCRIPTION
**Pull request by MarkRivers**
Add new ADSupport repository.  This repository contains:

- Source code versions of the hdf5, xml2, szip, and zlib libraries that were in ADBinaries
- Source code versions of the netCDF, Nexus, tiff, and jpeg libraries that were in ADCore

The change to ADCore is mostly to remove the source code for netCDF, Nexus, tiff, and jpeg.

The file plugin code, commonLibraryMakefile and commonDriverMakefile were also changed because jpeg and tiff are now supported on vxWorks, and there is new support for mingw.

### CODE REVIEW

**Comment by ulrikpedersen** [comment]
I will be reviewing this today and looking at the following:

- Building on our Linux setup.
- Making the travis build work again.
- Windows build (depending on time - may continue tomorrow).

Cheers,
Ulrik

**Comment by ulrikpedersen** [comment]
This is taking longer than expected to wrap my head around (and get working in our environment). I will have to continue tomorrow...

**Comment by ulrikpedersen** [comment]
I can build ADSupport and ADCore (no_ADBinaries branch) quite easily on our Linux system (RHEL6, gcc 4.4.7). The boost unittest system also run without errors.

The compilation of tiff in ADSupport failed for our VxWorks. Probably just due to our old VxWorks and/or gcc (2.96) versions. I wasn't expecting this to work.

It would be nice to have the available build options for each module (ADSupport and ADCore) documented in the top-level README.md in a standardised fashion. For example with a little table of the build system variables you can define in your CONFIG_SITE.arch.Common. This is of use to those of us who build each areaDetector module separately.

Our windows-x64 toolchain: MSVC 2010

The static build completes for both ADSupport and ADCore. I have not yet build/run an example IOC (TODO).

The shared build (which is how we normally build on Windows) fails in ADSupport when building NeXus.dll:

```
link /nologo /subsystem:windows /dll /LTCG /incremental:no /opt:ref /release   /MACHINE:X64     /implib:NeXus.lib /out:NeXus.dll         napi.obj napi5.obj napiu.obj nxdataset.obj nxio.obj nxstack.obj nxxml.obj stptok.obj       ..\\..\\..\\lib\\windows-x64\\szip.lib  ..\\..\\..\\lib\\windows-x64\\zlib.lib  ..\\..\\..\\lib\\windows-x64\\hdf5.lib 
   Creating library NeXus.lib and object NeXus.exp
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_SHORT_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT32_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT8_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UINT64_g
napi5.obj : error LNK2001: unresolved external symbol H5P_CLS_DATASET_CREATE_ID_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_USHORT_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UINT_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UCHAR_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_SCHAR_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_FLOAT_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UINT8_g
napi5.obj : error LNK2001: unresolved external symbol H5T_C_S1_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT16_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UINT16_g
napi5.obj : error LNK2001: unresolved external symbol H5P_CLS_FILE_ACCESS_ID_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_UINT32_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT64_g
napi5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_DOUBLE_g
NeXus.dll : fatal error LNK1120: 19 unresolved externals
make[3]: *** [NeXus.dll] Error 1120
```
See full build output in attached log file: 
[shared-build.log.tar.gz](https://github.com/areaDetector/ADCore/files/495971/shared-build.log.tar.gz)

**Still todo:**

- Figure out the Windows shared build problem - any ideas?
- Build and test example IOCs.
- Fix the travis build (should just be a matter of downloading and building ADSupport)

**Comment by MarkRivers** [comment]
> The shared build (which is how we normally build on Windows) fails in ADSupport when building NeXus.dll:

I rebuilt ADSupport for windows-x64 using MSVC 2010, generating a log file and comparing it to yours.  The major difference I can see is that when you compile C files it is using the /MT flag, while mine uses the /MD flag.  It should not use /MT if you are using the standard configuration files in base 3.14.12.5/configure/os/.  CONFIG.win32-x86.win32-x86 is what configures this, and it uses /MT for static builds and /MD for dynamic builds.  I suspect this is your problem, but I am not sure.

**Comment by MarkRivers** [comment]
> The compilation of tiff in ADSupport failed for our VxWorks. Probably just due to our old VxWorks and/or gcc (2.96) versions. I wasn't expecting this to work.

We are using vxWorks 6.9.2 which has gcc 4.3.3.  So I agree that is probably your problem.  If DLS plans to move to base 3.15 then I believe you will need to upgrade to vxWorks 6.x, at least according to the 3.15 Web page.

**Comment by MarkRivers** [comment]
> It would be nice to have the available build options for each module (ADSupport and ADCore) documented in the top-level README.md in a standardised fashion. For example with a little table of the build system variables you can define in your CONFIG_SITE.arch.Common. This is of use to those of us who build each areaDetector module separately.

I am not sure what you mean by the "build options".  Can you give a little example?

**Comment by ulrikpedersen** [comment]
> We are using vxWorks 6.9.2 which has gcc 4.3.3. So I agree that is probably your problem. If DLS plans to move to base 3.15 then I believe you will need to upgrade to vxWorks 6.x, at least according to the 3.15 Web page.

What I forgot to mention before is that I'm building on R3.14.12.3 as that is what we're using at Diamond. Eventually we will of course have to move on to 3.15 here, but we're not starting that just now.

**Comment by MarkRivers** [comment]
> What I forgot to mention before is that I'm building on R3.14.12.3 as that is what we're using at Diamond. Eventually we will of course have to move on to 3.15 here, but we're not starting that just now.

My tests are being run on 3.14.12.5, which is what we are using in production.

>It should not use /MT if you are using the standard configuration files in base 3.14.12.5/configure/os/. CONFIG.win32-x86.win32-x86 is what configures this, and it uses /MT for static builds and /MD for dynamic builds.

I just checked and 3.14.12.3 has the same settings as 3.14.12.5, i.e. use /MT for static builds and /MD for dynamic builds.  So it appears that you must have changed your settings somewhere since you are getting /MT for dynamic builds.

**Comment by ulrikpedersen** [comment]
> So it appears that you must have changed your settings somewhere since you are getting /MT for dynamic builds.

You're correct of course. For EPICS base we have set `STATIC_BUILD=YES` in `CONFIG_SITE.windows-x64.Common` which causes the `/MT` flags. However, we used to get away with building/linking against HDF5 as a shared library _somehow_. I'm a bit uneasy about jumping on to having a fully static build as we had never-ending trouble maintaining a complete static build of HDF5 in the past. I'm trying to figure out if I can have the option of a HDF5 shared dll somehow in combination with static ADCore and EPICS base. 

Its trial'n error, taking forever with our slow windows builds...

**Comment by ulrikpedersen** [comment]
> > It would be nice to have the available build options for each module (ADSupport and ADCore) documented in the top-level README.md in a standardised fashion. For example with a little table of the build system variables you can define in your CONFIG_SITE.arch.Common. This is of use to those of us who build each areaDetector module separately.

> I am not sure what you mean by the "build options". Can you give a little example?

Build options that are used for each module to make decisions in Makefiles in the individual module. For example for ADSupport you can set `SHARED_LIBRARIES` and/or `HDF5_STATIC_BUILD` - and ADCore has options to define BOOST...

I propose a small table on the top-level README.md can list the available options, valid values, etc. Could look something like this, along with an appropriate explanation about how to define these:

Build option        | Valid values        | Description
--------------------|---------------------|--------------------------------------------------------------
SHARED_LIBRARIES    | YES or NO           | Build libraries as shared (dynamic) or static
HDF5_STATIC_BUILD   | NO or undefined     | The HDF5 libraries are (or will be) built as static libraries

I can make a specific proposal and a separate PR.

**Comment by ulrikpedersen** [comment]
> I'm trying to figure out if I can have the option of a HDF5 shared dll somehow in combination with static ADCore and EPICS base.

Intermediate result: yes, I can do this (build succeeds, no test IOC yet). Just by setting:

```
SHARED_LIBRARIES=YES
HDF5_STATIC_BUILD=NO
```

In `ADSupport/configure/CONFIG_SITE.windows-x64.Common`

**Comment by MarkRivers** [comment]
Now that we are building HDF5 in ADSupport I would be interested to know if there you have any reason not to use the standard EPICS base configurations for Visual Studio:

win32-x86 (SHARED_LIBRARIES=YES, STATIC_BUILD=NO)
win32-x86-static (SHARED_LIBRARIES=NO, STATIC_BUILD=YES).

These are the only supported combinations of these parameters according to the EPICS documentation.

**Comment by MarkRivers** [comment]
> Its trial'n error, taking forever with our slow windows builds...

I have also suffered with slow Windows builds, but I just did some tests and found a way to speed them up by a factor of 6.  I have been doing my builds on a Linux file server called corvette.  I just did some benchmarks comparing building on a local Windows file system, on a remote Windows server file system, and on 3 different remote Linux file systems.  All builds were done using the same Windows workstation.

    File server           File server                     Connection      make –sj clean       make –sj8 
    computer name	computer                         type          (time in seconds)    (time in seconds)
    rivers-mobile2   Windows 7, 8 cores        Local hard drive      4                           75
	                      i7-2820QM,  2.30 GHz			
     cars5                Windows server               Network                16                         107
    bmw                  Linux, Fedora 23, 20 cores,  Network14	108
                              Xeon E5-2687W, 3.10 GHz	
    co7	Linux, RHEL7, 2 cores, Xeon 5150 2.66 GHz	Network	44	144
    corvette	Linux, Fedora 15, 8 cores,	Network	85	661
	Xeon E5630, 2.53 GHz

**Comment by ulrikpedersen** [comment]
> win32-x86 (SHARED_LIBRARIES=YES, STATIC_BUILD=NO)
win32-x86-static (SHARED_LIBRARIES=NO, STATIC_BUILD=YES).

> These are the only supported combinations of these parameters according to the EPICS documentation

As you know we don't use the top-level areaDetector module. Instead we configure and build each areaDetector repo as an individual/separate module. Our (windows) build configuration for each part is basically:

EPICS base (in `CONFIG_SITE.windows-x64.Common`):
```
SHARED_LIBRARIES=NO
STATIC_BUILD=YES
```

ADCore (in `CONFIG_SITE.windows-x64.Common`):
```
HDF5_STATIC_BUILD=NO
```

ADSupport (in `CONFIG_SITE.windows-x64.Common`):
```
SHARED_LIBRARIES=YES
HDF5_STATIC_BUILD=NO
```

In summary: EPICS base and ADCore is built statically, but for ADSupport we can override that and build DLLs out of the 3rd party libraries.

> Now that we are building HDF5 in ADSupport I would be interested to know if there you have any reason not to use the standard EPICS base configurations for Visual Studio

We are using a supported/standard build of EPICS base (see above). I am just interested in being able to build and link against 3rd party libs as DLLs. In the past I have had trouble maintaining a static build of HDF5 (and other 3rd party libs). Also, not all vendors ship static builds of their drivers...

**Comment by ulrikpedersen** [comment]
> I have also suffered with slow Windows builds, but I just did some tests and found a way to speed them up a lot. I have been doing my builds on a Linux file server called corvette. I just did some benchmarks comparing building on a local Windows file system, on a remote Windows server file system, and on 3 different remote Linux file server systems.

Yeah, this is a long issue that we've had here. The trouble is with our networked filesystem. Builds are snappy if **all** parts (toolchain + all dependencies) reside on a local disk. I've set it up a couple of times but I find it's too much effort to maintain over longer periods. So then I just suffer the wait. I hear that we're about to replace our +6 year old file system with a new, and hopefully snappier one.

**Comment by MarkRivers** [comment]
> In summary: EPICS base and ADCore is built statically, but for ADSupport we can override that and build DLLs out of the 3rd party libraries.
When you do this do you end up with some code compiled with /MT and some code compiled with /MD?  If so then do you have a problem with the system being confused about whether to link with MSVCRT.lib or LIBCMT.lib?

**Comment by ulrikpedersen** [comment]
> When you do this do you end up with some code compiled with /MT and some code compiled with /MD? If so then do you have a problem with the system being confused about whether to link with MSVCRT.lib or LIBCMT.lib?

I agree, that this is a bit funny and I should have an issue, but it doesn't appear to be - not previously and not now. I'll try to rebuild and log (and time) each of my ADSupport and ADCore windows builds for investigation.

**Comment by MarkRivers** [comment]
Note that in commonDriverMakefile and commonLibraryMakefile I have removed all occurrences of the /NODEFAULTLIB flag.  These were previously necessary when building statically and not using GraphicsMagick, or when building the debug version.  This is because the vendor DLLs were built with different flags.  Since we are no longer linking with any pre-built vendor libraries in ADCore or ADExample the NODEFAULTLIB flags are no longer needed.  They may be needed in specific detector modules if the vendor library was built with different flags, and may need to be added to those individual Makefiles.  However, I not yet seen any errors when building the detector modules.

Note also that GraphicsMagic is currently unsupported on Windows.  I plan to support it in the future by building it from source in ADSupport, just like the other 3rd party libraries.  But I may not do this before the R2-5 release.

I am also working to change the build system in ADSupport and ADCore to that each of the 3rd party libraries has the following flags, using HDF5 as an example:

WITH_HDF5=[YES,NO]  If this flag is NO then ADSupport and ADCore do not build any code that requires HDF5.

HDF5=[path to external HDF library]  If this is undefined then the code in ADSupport is built and used.  If it is defined then the library is assumed to be present somewhere external to areaDetector.

HDF5_INCLUDE=[path to external HDF5 include files]
HDF5_DIR=[path to external HDF5 libraries]

We may not need both HDF5 and HDF5_DIR. I need to test this.

I also plan to add the files required to build all of ADSupport on Linux.  There have been several requests to allow building HDF5 in ADSupport on Linux.  This is easy to do, using the 32-bit and 64-bit files generated by "configure" on my system.  Users who want to do this may need to edit these files if there are building on a different Linux architecture (ARM, etc.) or a Linux installation that has something different from our Linux system.

**Comment by ulrikpedersen** [comment]
> I am also working to change the build system in ADSupport and ADCore to that each of the 3rd party libraries has the following flags, using HDF5 as an example:
> 
> WITH_HDF5=[YES,NO] If this flag is NO then ADSupport and ADCore do not build any code that requires HDF5.
> 
> HDF5=[path to external HDF library] If this is undefined then the code in ADSupport is built and used. If it is defined then the library is assumed to be present somewhere external to areaDetector.
> 
> HDF5_INCLUDE=[path to external HDF5 include files]
> HDF5_DIR=[path to external HDF5 libraries]

Perfect! 

> We may not need both HDF5 and HDF5_DIR. I need to test this.

You may need to make some assumptions about where libs and include files have been installed. That may not be the same on all platforms...

**Comment by ulrikpedersen** [comment]
For information:

I have now done another windows-x64 build of ADCore, ADSupport and simDetectorIOC (from ADExample). The build configurations is as mentioned previously:

EPICS base (in `CONFIG_SITE.windows-x64.Common`):
```
SHARED_LIBRARIES=NO
STATIC_BUILD=YES
```

ADCore (in `CONFIG_SITE.windows-x64.Common`):
```
HDF5_STATIC_BUILD=NO
```

ADSupport (in `CONFIG_SITE.windows-x64.Common`):
```
SHARED_LIBRARIES=YES
HDF5_STATIC_BUILD=NO
```

simDetectorIOC (in `CONFIG_SITE.windows-x64.Common`):
```
HDF5_STATIC_BUILD=NO
```

The build logs are captured here: 
[static-build-dynamic-adsupport.logs.tar.gz](https://github.com/areaDetector/ADCore/files/502831/static-build-dynamic-adsupport.logs.tar.gz)

I've grepped through the logs and find that only the `/MT` and `-DH5_BUILT_AS_DYNAMIC_LIB` has been used. The `/MD` and `-DH5_BUILT_AS_STATIC_LIB` are nowhere to be found. I have ended up with DLLs and lib stub files in ADSupport:

```
(master) [up45@pc0009 ADSupport]$ ls -lh bin/windows-x64/
total 8.0M
-rwxrwxr-x. 1 up45 up45 161K Sep 29 14:01 H5detect.exe
-rwxrwxr-x. 1 up45 up45 4.2M Sep 29 14:01 hdf5.dll
-rwxrwxr-x. 1 up45 up45 219K Sep 29 13:47 jpeg.dll
-rwxrwxr-x. 1 up45 up45 839K Sep 29 13:46 netCDF.dll
-rwxrwxr-x. 1 up45 up45 158K Sep 29 14:01 NeXus.dll
-rwxrwxr-x. 1 up45 up45  92K Sep 29 13:50 szip.dll
-rwxrwxr-x. 1 up45 up45  51K Sep 29 13:46 test_big_classic.exe
-rwxrwxr-x. 1 up45 up45 528K Sep 29 13:48 tiff.dll
-rwxrwxr-x. 1 up45 up45 1.6M Sep 29 13:49 xml2.dll
-rwxrwxr-x. 1 up45 up45 136K Sep 29 13:50 zlib.dll
(master) [up45@pc0009 ADSupport]$ ls -lh lib/windows-x64/
total 1.1M
-rwxrwxr-x. 1 up45 up45 551K Sep 29 13:59 hdf5.lib
-rwxrwxr-x. 1 up45 up45  23K Sep 29 13:47 jpeg.lib
-rwxrwxr-x. 1 up45 up45  50K Sep 29 13:46 netCDF.lib
-rwxrwxr-x. 1 up45 up45  15K Sep 29 14:01 NeXus.lib
-rwxrwxr-x. 1 up45 up45 3.3K Sep 29 13:50 szip.lib
-rwxrwxr-x. 1 up45 up45  36K Sep 29 13:47 tiff.lib
-rwxrwxr-x. 1 up45 up45 337K Sep 29 13:49 xml2.lib
-rwxrwxr-x. 1 up45 up45  15K Sep 29 13:50 zlib.lib
```

This works for me and I am pleased that I can build/use DLLs in combination with a static build - at least if I control the DLL build process! With 3rd party pre-built libs/dlls there will always be this conflict between `/MD` and `/MT` of course...

**Comment by MarkRivers** [comment]
Yes, I am working on other changes.  See my comment this morning in #190.  I am currently working on the no_ADBinaries branch of areaDetector and the master branch of ADSupport.  It is not complete yet.  It will also require changes to commonDriverMakefile and commonLibraryMakefile in ADCore, so let's postpone the merge or ADCore  until that is working.

**Comment by ulrikpedersen** [comment]
> Yes, I am working on other changes. See my comment this morning in #190. I am currently working on the no_ADBinaries branch of areaDetector and the master branch of ADSupport. It is not complete yet. It will also require changes to commonDriverMakefile and commonLibraryMakefile in ADCore, so let's postpone the merge or ADCore until that is working.

Sure! That's OK with me.

**Comment by MarkRivers** [comment]
I have finished working in areaDetector (top-level), ADSupport, and ADCore to make the libraries and plugins for things like TIFF, netCDF, HDF5, etc. optional.  I have tested building using WITH_XXX=YES for everything except XXX=GraphicsMagick.  On Linux I used XXX_EXTERNAL=YES for everything except NETCDF and NEXUS.  On other architectures (win32-x86, win32-x86-static, windows-64, windows-x64-static, win32-x86-mingw, win32-x86-mingw-static, vxWorks-ppc32) I used XXX_EXTERNAL=NO for everything.

I still need to test XXX_EXTERNAL=NO for all XXX on Linux.

I am now ready for the merge, and then for people to test the master branch of ADCore and areaDetector.

**Comment by MarkRivers** [comment]
> I still need to test XXX_EXTERNAL=NO for all XXX on Linux.

I have now tested using XXX_EXTERNAL=NO for linux-x86 and linux-x86_64.  This means it was using the versions of the hdf5, szip, xml2, and zlib libraries from ADSupport rather than the external or system versions of these libraries.

**Comment by MarkRivers** [comment]
@ulrikpedersen OK to merge.
